hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -363,118 +403,89 @@ protected void stopNetworking() {
+    public void initialConnectToCluster() {
         clusterDiscoveryService.current().start();
-
         if (asyncStart) {
-            submitConnectToClusterTask();
+            asyncConnectToCluster();
         } else {
-            doConnectToCluster();
+            connectToCluster();
         }
+        clusterExecutor.execute(this::clusterThreadLoop);
     }
 
-    private void submitConnectToClusterTask() {
-        // called in synchronized(clusterStateMutex)
+    private void asyncConnectToCluster() {
+        taskQueue.add(this::retryConnectToCluster);
+    }
 
-        if (connectToClusterTaskSubmitted) {
-            return;
+    private void retryConnectToCluster() {
+        if (!client.getLifecycleService().isRunning()) {
+            shutdownWithExternalThread();
+        }
+        try {
+            connectToCluster();
+        } catch (Throwable e) {
+            logger.warning("Could not connect to any cluster, shutting down the client: " + e.getMessage());
+            shutdownWithExternalThread();
         }
-
-        executor.submit(() -> {
-            try {
-                doConnectToCluster();
-                synchronized (clientStateMutex) {
-                    connectToClusterTaskSubmitted = false;
-                    if (activeConnections.isEmpty()) {
-                        if (logger.isFineEnabled()) {
-                            logger.warning("No connection to cluster: " + clusterId);
-                        }
-
-                        submitConnectToClusterTask();
-                    }
-                }
-            } catch (Throwable e) {
-                logger.warning("Could not connect to any cluster, shutting down the client: " + e.getMessage());
-                shutdownWithExternalThread();
-            }
-        });
-
-        connectToClusterTaskSubmitted = true;
     }
 
-    private void doConnectToCluster() {
+    /**
+     * If this throws exception it means instance is either not started or will be shutdown.
+     */
+    private void connectToCluster() {
         CandidateClusterContext currentContext = clusterDiscoveryService.current();
-
         logger.info("Trying to connect to cluster: " + currentContext.getClusterName());
-
         // try the current cluster
-        if (doConnectToCandidateCluster(currentContext, false)) {
+        AuthenticationResult result = doConnectToCandidateCluster(currentContext);
+        if (result != null && checkClusterId(result) && onAuthenticatedToCluster(result)) {
             return;
         }
-
-        synchronized (clientStateMutex) {
-            if (activeConnections.isEmpty()) {
-                clientState = ClientState.SWITCHING_CLUSTER;
-            } else {
-                //ConnectToAllClusterMembersTask connected back to the same cluster
-                //we don't need to switch cluster anymore.
-                return;
-            }
-        }
-
+        clientState = ClientState.SWITCHING_CLUSTER;
```

Comment:
I don't like the fact that we are setting this, regardless of the blue/green enabled. We might reach this line easily without blue/green and I am not %100 sure it is safe to set state to this for others part of the code

Reply:
After simplifying the states this turned out to be not necessary. See 
https://github.com/hazelcast/hazelcast/commit/97b786836c2085b5ec827c0a93b3d91dd29a4089